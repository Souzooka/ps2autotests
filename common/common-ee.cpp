#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
