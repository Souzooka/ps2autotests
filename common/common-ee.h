#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
