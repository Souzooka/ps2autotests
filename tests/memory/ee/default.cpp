#include <common-ee.h>
#include <malloc.h>

static const char *describePtr(void *p) {
	u32 v = (u32)p;
	if (v >= 0x80000000 && v < 0x80100000) {
		return "EE Kernel";
	}
	if (v >= 0x80100000 && v < 0x82000000) {
		return "EE KSEG0";
	}
	if (v >= 0xA0100000 && v < 0xA2000000) {
		return "EE KSEG1";
	}
	if (v >= 0x00100000 && v < 0x02000000) {
		return "EE Cached";
	}
	if (v >= 0x10000000 && v < 0x14000000) {
		return "HW Registers";
	}
	if (v >= 0x20100000 && v < 0x22000000) {
		return "EE Uncached";
	}
	if (v >= 0x30100000 && v < 0x32000000) {
		return "EE UCAB";
	}
	if (v >= 0xBC000000 && v < 0xBC200000) {
		return "IOP";
	}
	if (v >= 0x70000000 && v < 0x70004000) {
		return "Scratchpad";
	}
	return "Other";
}

static void tryAlloc(u32 mb) {
	void *p = malloc(mb * 1024 * 1024);
	printf("  %d MB: %s\n", mb, p ? "success" : "fail");
	free(p);
}

static void verifyMirrors() {
	void *p = calloc(1, 1024 * 1024);

	printf("  Allocated in %s\n", describePtr(p));
	printf("  Pointer is %s\n", ((u32)p & 0xF) == 0 ? "aligned" : "not aligned");

	u8 *cached = (u8 *)((u32)p & 0x00FFFFFF);
	u8 *uncached = (u8 *)((u32)cached | 0x20000000);
	u8 *ucab = (u8 *)((u32)cached | 0x30000000);

	*uncached = 42;

	printf("  Mirrors: cached=%d, uncached=%d, ucab=%d\n", *cached, *uncached, *ucab);

	free(p);
}

void verifyScratchpad() {
	// TODO: Is there any system to allocate the scratchpad?  Thread flag?

	u8 *sram = (u8 *)0x70000000;

	sram[0x0000] = 42;
	sram[0x3FFF] = 1;

	printf("  Scratchpad start: %d, end: %d\n", sram[0x0000], sram[0x3FFF]);
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	printf("Memory allocation limits:\n");
	// Let's not be to exact, since the program size or other things may make this inexact.
	tryAlloc(32);
	tryAlloc(30);
	tryAlloc(24);
	tryAlloc(16);

	printf("Mirrors:\n");
	verifyMirrors();

	printf("Scratchpad:\n");
	verifyScratchpad();

	printf("-- TEST END\n");
	return 0;
}