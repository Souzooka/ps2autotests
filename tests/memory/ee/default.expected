-- TEST BEGIN
Memory allocation limits:
  32 MB: fail
  30 MB: success
  24 MB: success
  16 MB: success
Mirrors:
  Allocated in EE Cached
  Pointer is aligned
  Mirrors: cached=42, uncached=42, ucab=42
Scratchpad:
  Scratchpad start: 42, end: 1
-- TEST END
